treescript/mercurial.py

```diff
@@ -0,0 +1,196 @@
+
+
+async def checkout_repo(context, directory):
+    """Perform a clone via robustcheckout, at ${directory}/src.
+
+    This function will perform a clone via robustcheckout, using hg's share extension
+    for a cache at `context.config['hg_share_base_dir']` that robustcheckout will
+    populate if necessary.
+
+    Robustcheckout will retry network operations at most 3 times (robustcheckout's default)
+    before giving up and causing FailedSubprocess to be raised.
+
+    Args:
+        context (TreeScriptContext): the treescript context
+        directory (str): The directory to place the resulting clone.
+
+    Raises:
+        FailedSubprocess: if the clone attempt doesn't succeed.
+
+    """
+    share_base = context.config['hg_share_base_dir']
+    upstream_repo = context.config['upstream_repo']
+    dest_repo = get_source_repo(context.task)
+    dest_folder = os.path.join(directory, 'src')
+    # branch default is used to pull tip of the repo at checkout time
+    await run_hg_command(context, 'robustcheckout', dest_repo, dest_folder,
+                         '--sharebase', share_base,
+                         '--upstream', upstream_repo,
+                         '--branch', 'default')
+
+
+async def do_tagging(context, directory):
+    """Perform tagging, at ${directory}/src.
+
+    This function will perform a mercurial tag, on 'default' head of target repository.
+    It will tag the revision specified in the tag_info portion of the task payload, using
+    the specified tags in that payload.
+
+    Tags are forced to be created at the specified revision if they already existed.
+
+    This function has the side affect of pulling the specified revision from the
+    destination repository. This feature exists because mozilla-unified does not
+    contain relbranches, though some releases are created on relbranches, so we must ensure
+    the desired revision to tag is known to the local repository.
+
+    Args:
+        context (TreeScriptContext): the treescript context
+        directory (str): The directory to place the resulting clone.
+
+    Raises:
+        FailedSubprocess: if the tag attempt doesn't succeed.
+
+    """
+    local_repo = os.path.join(directory, 'src')
+    tag_info = get_tag_info(context.task)
+    desired_tags = tag_info['tags']
+    desired_rev = tag_info['revision']
+    dest_repo = get_source_repo(context.task)
+    commit_msg = TAG_MSG.format(revision=desired_rev, tags=', '.join(desired_tags))
+    log.info("Pulling {revision} from {repo} explicitly.".format(
+        revision=desired_rev, repo=dest_repo))
+    await run_hg_command(context, 'pull', '--revision', desired_rev, dest_repo,
+                         repo_folder=local_repo)
+    log.info(commit_msg)
+    await run_hg_command(context, 'tag', '-m', commit_msg, '-r', desired_rev,
+                         '-f',  # Todo only force if needed
+                         *desired_tags,
+                         repo_folder=local_repo)
```

Comment:
I think we're missing a commit and push/rebase or push/strip loop.

Reply:
That is a todo, early blocker but manual tc rerun should be relatively simple in short-term